Fix collision in LabelCategories._indices

Bobronium opened the pull request:
<!-- Contributing guide: https://github.com/cvat-ai/datumaro#contributing -->

### Summary
Now indices keys in LabelCategories are computed as (parent or "") + name.
This may result in a collision with
```python
name1 = "22"
parent1 = "parent"
name2 = "parent2"
name2 = "2"

parent1 + name1 == parent2 + name2
```
So this code would fail:
```python
categories = LabelCategories()
categories.add("22", parent="parent")
categories.add("2", parent="parent2")  # AssertionError
```

Changing keys formula to `(parent, name)` (`tuple[str, str]`) solves the issue

### How to test
Changes include `tests.test_label_categories.test_potential_collision` that tests this exact behaviour.

### Checklist
<!-- Put an 'x' in all the boxes that apply -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/cvat-ai/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](https://github.com/cvat-ai/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ x I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 CVAT.ai Corporation
#
# SPDX-License-Identifier: MIT
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `labels` property in the `Category` class for easier access to label information.
  
- **Improvements**
	- Enhanced the clarity and organization of label indexing in the `Category` class.
	- Simplified label retrieval in the `camvid_format` plugin, improving readability and potentially performance.
	- Clarified the logic for determining `sublabel` in the `coco_format` extractor.

- **Bug Fixes**
	- Improved error message formatting in the `compare_datasets` function for better readability during test failures.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Bobronium committed:
Fix collision in LabelCategories._indices

Bobronium committed:
Fix formatting

zhiltsov-max reviewed (commented):
Could you add a test that reproduces the collision possible previously? Please also add some PR notes and a changelog line.

zhiltsov-max commented:
Please update license headers in modified files as well, where necessary.

Bobronium committed:
Fix LabelCategories.items type annotation

Bobronium committed:
Introduce conftest.ASSETS_DIR for consistent assets access

Bobronium committed:
Add tests for label categories

Bobronium committed:
Fix linting issues

Bobronium committed:
Update copyright headers

Bobronium committed:
Update CHANGELOG.md

zhiltsov-max commented on tests/test_label_categories.py:
```diff
@@ -0,0 +1,128 @@
+def test_add_duplicate_category():
+    categories = LabelCategories()
+    categories.add("cat")
+    with pytest.raises(KeyError):
```
It makes sense to check the message here, as KeyError is quite a broad thing, which can come from different sources.

Bobronium replied:
Agreed

zhiltsov-max commented on tests/test_label_categories.py:
```diff
@@ -0,0 +1,128 @@
+        categories.add("cat")
+
+
+def test_potential_collision():
+    """
+    Previously indices were computed as (parent or "") + name
+    """
```
It's better to change the description to include information about the fixing PR and the problem, as this test checks for regressions in a specific use case. Typically, it would also include a link to the issue, but we don't have one in this case.

Bobronium replied:
Included a link to the PR

zhiltsov-max commented on tests/test_label_categories.py:
```diff
@@ -0,0 +1,128 @@
+from datumaro import LabelCategories
+
+
+def test_add_category():
```
Please add test decorators, as in all other tests.

Bobronium replied:
Done!

Bobronium committed:
Add link to related PR

Bobronium committed:
Test KeyError message

zhiltsov-max commented on tests/test_label_categories.py:
```diff
@@ -0,0 +1,134 @@
+
+def test_add_duplicate_category():
+    categories = LabelCategories()
+    categories.add("cat")
+    with pytest.raises(KeyError, match="Label '' 'cat' already exists"):
+        categories.add("cat")
+
+    categories.add("cat", parent="animal")
+    with pytest.raises(KeyError, match="Label 'animal' 'cat' already exists"):
+        categories.add("cat", parent="animal")
```
Mixing several test cases in a single test call is a bad practice. Please consider using `@pytest.mark.parametrize` or make 2 tests.

Bobronium replied:
Good point, thanks!

Bobronium committed:
Add @mark_requirement(Requirements.DATUM_GENERAL_REQ) and parametrize tests

Bobronium committed:
Comply with isort

Bobronium committed:
Add licence header to test_label_categories.py

Bobronium committed:
Comply with isort

zhiltsov-max commented on tests/test_label_categories.py:
```diff
@@ -0,0 +1,152 @@
+
+
+@mark_requirement(Requirements.DATUM_GENERAL_REQ)
+def test_add_category_with_parent():
```
Please rename tests to use naming convention: `test_can_...`

Bobronium replied:
Done!

zhiltsov-max reviewed (approved):
Please update test names, otherwise lgtm.

Bobronium committed:
Rename tests to comply with current naming convention

Bobronium committed:
Add meaningful test for _reindex, fix _reindex

Bobronium commented:
Pushed another commit with changes to `_reindex` and its tests.

zhiltsov-max commented on datumaro/components/annotation.py:
```diff
@@ -147,13 +147,12 @@ def __attrs_post_init__(self):
 
     def _reindex(self):
-        indices = {}
+        self._indices = indices = {}
```
Please avoid using multiple assignment patterns in a single line of code. There is nothing wrong about it, it just makes the code harder to read and/or understand. It's related to patterns like this one and also `a, b = b, a`, `a = 0; b = 1; c = 2`.

Bobronium committed:
Remove alias for self._indices

Not entirely agree with prohibition of chained assignments, but in this case it just might be unnecessary.

nmanovic commented:
I'm not sure, but probably you need to apply recommendations from the issue: https://github.com/pyinstaller/pyinstaller/issues/8713

zhiltsov-max commented on tests/test_mars_format.py:
```diff
@@ -14,10 +13,10 @@
 
-ASSETS_DIR = osp.join(osp.dirname(__file__), "assets")
-DUMMY_MARS_DATASET = osp.join(ASSETS_DIR, "mars_dataset")
+DUMMY_MARS_DATASET = str(ASSETS_DIR / "mars_dataset")
```
```suggestion
DUMMY_MARS_DATASET = os.fspath(ASSETS_DIR / "mars_dataset")
```
In other places as well.

Bobronium replied:
`str(Path(...))` should be sufficient here, IMO. What prompted the suggestion?

If there's a strong need to forcefully resolve the path (which there isn't since `ASSETS_DIR` is an absolute path), I would go with `str((ASSETS_DIR / "mars_dataset").resolve())`, yet again I consider that or `os.fspath(...)` redundant code in this context.

My suggestion is to leave this as is.

zhiltsov-max replied:
If the value is supposed to be used as a path in system calls, the `os.fspath` is the correct function to get a string representation.

Bobronium replied:
The only utility for `os.fspath` is to be used to get string representation of an arbitrary `PathLike` object. 

Here we're dealing not with arbitrary `PathLike` object, but with `pathlib.PurePath` subclass. 

A canonical way to get `PurePath` string representation is using `str(path)`.
